### PR TITLE
Add a PhoneNumber class to handle formatting and actions on numbers

### PR DESCRIPTION
**Pull request by caitlinrhd**
We value collaboration at Invoca, and focus on peer mentorship as a way for the whole team to learn and thrive. As part of this process, we conduct code reviews during our development cycle. In this phase of the interview, we would like you to review and comment on this Pull Request. 

You can assume that this Pull Request was written by an intern, who is new to professional software development.  Please take about 20-30 minutes to review this Pull Request and comment as if you were reviewing an intern's code.

### CODE REVIEW

File: phone_number.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    "(#{find_area_code}) #{find_exchange}-#{find_line_number}"
+  end
+
+  def find_area_code
+    open_paren = @number.index('(')
+    close_paren = @number.index(')')
+    if open_paren
+      if close_paren
+        @number[(open_paren + 1)..(close_paren - 1)]
+      end
```
Comment (nitishmadhukar) [inline phone_number.rb:46]: This is a good check to verify the existence of the `( )` parenthesis. But, we are missing some edge cases like the input phone number having only the open parenthesis but no closing parenthesis. These edge cases might not occur if the phone number is sent by a controlled system. But if it is a free form user field, then this check fails. The below steps help in making the check more robust.

1. We could sanitize the input and not allow any characters other than numbers
2. We could allow parenthesis but check if both opening and closing parenthesis are included during validation. In that case, we could use the below snippet to simplify the conditionals
```
if open_paren && close_paren
  .......
else
  .......
```

File: phone_number_spec.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+require_relative "./phone_number"
+
+RSpec.describe PhoneNumber do
+  let(:phone_number) { PhoneNumber.new("5558675309") }
+
+  describe "#number" do
+    it "is a formatted number" do
```
Comment (nitishmadhukar) [inline phone_number_spec.rb:7]: This is a good positive test scenario

File: phone_number_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      expect(phone_number.number).to eq("(555) 867-5309")
+    end
+  end
+
```
Comment (nitishmadhukar) [inline phone_number_spec.rb:11]: Could you also include some tests to test the negative scenarios.  

We can include some tests having phone numbers with parenthesis and phone numbers with less than 10 digits and see how the system is responding.

File: phone_number_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  describe "#text" do
+    it "calls PhoneActions#text!" do
+      expect_any_instance_of(PhoneActions).to receive(:text!)
+      phone_number.text("hello world!")
+    end
+  end
+
+  describe "#call" do
+    it "calls PhoneActions#call!" do
+      expect_any_instance_of(PhoneActions).to receive(:call!)
```
Comment (nitishmadhukar) [inline phone_number_spec.rb:28]: Same as above, we can extend `expect_any_instance_of` to use `with(<args>)`

**Comment by nitishmadhukar** [comment]
Appreciate having all the tests in green before submitting the PR. It is a good start. 

It is good to have many unit tests as they run fast and provide quick feedback. Every method should have at least one unit test. But when there is an external dependency, an integration/service test offers more value.  

In this PR, as we stubbed out the entire method that calls the external API dependency (`call!` and `text!`), we were not verifying if the `PhoneNumber` class is able to call the API inside `PhoneActions`. A better approach would be to just stub the HTTP request(RestClient call) and have an integration spec to call the actual method inside `PhoneActions`.    A great read to understand the various levels of testing is https://martinfowler.com/articles/practical-test-pyramid.html#TheTestPyramid

**Comment by nitishmadhukar** [comment]
Once you have the `PhoneActions` class updated to the `API::Phone` class, can you please add unit tests to the `post` method. The tests will help us to verify the `API::Phone` and `PhoneNumber` classes in isolation. 

An integration test will call the method without stubbing which will help us verify the entire flow.

FYI, some external APIs have rate limits, so we need to run the integration specs conservatively.
